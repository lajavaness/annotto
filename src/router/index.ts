import express from 'express'
import control from '../middleware/control'
import addUsersRoutes from './users'
import addProjectsRoutes from './projects'
import addItemsRoutes from './items'
import addClientsRoutes from './clients'
import addFilterRoutes from './filters'
import addProfilesRoutes from './profiles'
import { keycloak } from '../auth'

const apiRouter = () => {
  const r = express.Router()

  /**
   * Get health status.
   * @route GET /health
   * @group health-api
   * @security Bearer
   * @returns {Health.model} 200 - Success.
   * @returns {AnnottoError.model} 400 - Invalid Credentials Error.
   * @returns {AnnottoError.model} 500 - Place get error.
   */
  r.get('/health', control.health)

  // After this line, every routes are keycloak dependant.
  r.use(keycloak.middleware())

  addUsersRoutes(r)
  addProjectsRoutes(r)
  addItemsRoutes(r)
  addClientsRoutes(r)
  addFilterRoutes(r)
  addProfilesRoutes(r)

  /**
   * @typedef List
   * @property {number} count Page length ( to set page max length to other value than default 10, use limit query param ).
   * @property {number} index Index of the page ( to access specific index use index query param ).
   * @property {number} pageCount Number of Page ( data length / page length ).
   * @property {number} total Total data length.
   * @property {Array.<object>} data
   */

  /**
   * @typedef UserExtern
   * @property {string} _id _id.
   * @property {string} firstName FirstName of the item.
   * @property {string} lastName LastName of the user.
   * @property {string} email Email.
   */

  /**
   * @typedef AnnottoError
   * @property {number} code Error code - eg: 401.
   * @property {string} message - Error message.
   */

  /**
   * @typedef Health
   * @property {string} status Status of the server.
   * @property {string} version Current version of the API.
   * @property {string} environment Current environment of the API.
   */

  /**
   * @typedef User
   * @property {string} firstName FirstName of the user.
   * @property {string} lastName LastName of the user.
   * @property {number} userId Id of the user.
   * @property {boolean} isActive User can login or not.
   * @property {string} email Email address of the user.
   */

  /**
   * @typedef UserUpdate
   * @property {string} firstName FirstName of the user.
   * @property {string} lastName LastName of the user.
   */

  /**
   * @typedef Auth
   * @property{string} email
   * @property {string} password Password.
   */

  /**
   * @typedef LoginCred
   * @property {number} userId Id of the user.
   * @property {string} token Token to provide.
   */

  return r
}

export default apiRouter
