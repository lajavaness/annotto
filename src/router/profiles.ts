import { Router } from 'express'
import auth from '../auth/service'
import profiles from '../middleware/profiles'
import { schemaValidationMiddleware } from '../middleware/schema-validation-middleware'
import { rolesSchema } from './validation/profile'

/**
 * @typedef ProfilesList
 * @property {number} count - Number of profiles.
 * @property {number} index - Index of the page.
 * @property {number} pageCount - Page quantity.
 * @property {number} total - Total number of profiles.
 * @property {Array.<Profile>} data - List of profiles.
 */

/**
 * @typedef Profile
 * @property {string} user.required - Profile user-api ID.
 * @property {boolean} role.required - Profile role.
 * @property {string} email.required - Profile email.
 */

const init = (router: Router) => {
  /**
   * List Profiles
   * See confluence Doc for more details : https://lajavaness.atlassian.net/wiki/spaces/AN/pages/2086371385/Permissions.
   * @route GET /profiles
   * @group profiles
   * @security Bearer
   * @returns {ProfilesList.model} 200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 Fail to list profiles.
   */
  router.get('/profiles', auth.isAuthenticated({ realmRole: 'admin' }), profiles.index)

  /**
   * Update Profile role.
   * @route PUT /profiles/:id
   * @group profiles
   * @security Bearer
   * @param {string} role.query.required Query param "role" to assign to user ( "user / dataScientist / admin" ).
   * @returns {ProfilesList.model} 200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 Fail to list profiles.
   */
  router.put(
    '/profiles/:profileId',
    auth.isAuthenticated({ realmRole: 'admin' }),
    schemaValidationMiddleware({ schemaQuery: rolesSchema }),
    profiles.update
  )
}

export default init
