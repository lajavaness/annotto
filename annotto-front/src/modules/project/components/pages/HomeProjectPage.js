import { Button, Dropdown, Menu, Modal } from 'antd'
import { ExclamationCircleOutlined } from '@ant-design/icons'
import { isEmpty, isNumber } from 'lodash'
import { navigate } from '@reach/router'
import { useDispatch, useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import React, { useCallback, useEffect, useMemo, useState } from 'react'
import dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime'

import {
	deleteProject,
	fetchProjectItems,
	fetchProjectItemsAndLogs,
	fetchProjectLogs,
	fetchProjectStatsTasks,
	navigateToItem,
	openFilterModal,
	openGuideModal,
} from 'modules/project/actions/projectActions'

import {
	selectProjectAnnotators,
	selectProjectFilter,
	selectProjectId,
	selectProjectItems,
	selectProjectItemsIndex,
	selectProjectItemsTotal,
	selectProjectLogs,
	selectProjectLogsIndex,
	selectProjectLogsTotal,
	selectProjectStatsTasks,
	selectProjectStatsTasksIndex,
	selectProjectStatsTasksTotal,
} from 'modules/project/selectors/projectSelectors'
import { selectUserInfo } from 'modules/root/selectors/authSelectors'

import { DELETE_PROJECT, EDIT_PROJECT, EDIT_PROJECT_EXPORT } from 'shared/enums/privilegesTypes'
import { ITEMS } from 'shared/enums/projectStatsTypes'
import { PROJECT_ITEMS_SIZE, PROJECT_LOGS_SIZE, PROJECT_STATS_TASKS_SIZE } from 'shared/enums/paginationTypes'

import useProjectPrivileges from 'shared/hooks/useProjectPrivileges'

import AnnotatorsContainer from 'modules/project/components/common/AnnotatorsContainer'
import Card from 'shared/components/card/Card'
import Col from 'shared/components/grid/Col'
import LogsContainer from 'modules/project/components/common/LogsContainer'
import Page from 'shared/components/page/Page'
import Row from 'shared/components/grid/Row'
import WarningMessage from 'shared/components/warning/WarningMessage'

import { groupTasksByCategories } from 'shared/utils/tasksUtils'
import { isHTML } from 'shared/utils/htmlUtils'

import * as Styled from 'modules/project/components/pages/__styles__/HomeProjectPage.styles'
import { TASKS } from 'shared/enums/projectStatsTypes'

// eslint-disable-next-line react/prop-types
const HomeProjectPage = ({ setHeaderActions }) => {
	const [selectedStatsDisplay, setSelectedStatsDisplay] = useState(ITEMS)

	const projectId = useSelector(selectProjectId)
	const filter = useSelector(selectProjectFilter)
	const items = useSelector(selectProjectItems)
	const itemsIndex = useSelector(selectProjectItemsIndex)
	const totalItems = useSelector(selectProjectItemsTotal)
	const tasks = useSelector(selectProjectStatsTasks)
	const tasksIndex = useSelector(selectProjectStatsTasksIndex)
	const totalTasks = useSelector(selectProjectStatsTasksTotal)
	const logs = useSelector(selectProjectLogs)
	const logsIndex = useSelector(selectProjectLogsIndex)
	const totalLogs = useSelector(selectProjectLogsTotal)
	const annotators = useSelector(selectProjectAnnotators)
	const currentUser = useSelector(selectUserInfo)

	const dispatch = useDispatch()

	const { t } = useTranslation(['root', 'project'])

	const hasPrivilege = useProjectPrivileges()

	useEffect(() => {
		/**
		 * We dispatch the action to trigger the fetch of items and logs here rather than in the startup saga.
		 * Because we need to retrieve this data each time this page is mounted and not when the module is mounted.
		 * When we are on the annotation page and when we navigate to this component,
		 * this data is not fetched since we are in the same module and the startup saga is not triggered.
		 */
		dispatch(fetchProjectItemsAndLogs(projectId))
	}, [projectId, dispatch])

	dayjs.extend(relativeTime)

	const itemsData = useMemo(
		() =>
			items?.map(({ _id, body, logCount, velocity, annotated, annotationValues, annotatedAt }, index) => ({
				key: index,
				body,
				logCount,
				velocity,
				annotated,
				annotationValues,
				annotatedAt,
				_id,
			})) || [],
		[items]
	)

	const resolveTasksByCategories = useMemo(
		() =>
			groupTasksByCategories(tasks).map(({ category, children, ...others }, index) => ({
				key: index,
				value: category,
				children: children.map((child, index) => ({ key: `category_child_${index}`, ...child })),
			})),
		[tasks]
	)

	const _onEditProjectClick = useCallback(
		(e) => {
			e.stopPropagation()
			return navigate(`/edit_project/${projectId}/config`)
		},
		[projectId]
	)

	const _onExportProjectClick = useCallback(
		(e) => {
			e.stopPropagation()
			return navigate(`/edit_project/${projectId}/export`)
		},
		[projectId]
	)

	const _onDeleteProjectClick = useCallback(
		(e) => {
			e.stopPropagation()
			Modal.confirm({
				icon: <ExclamationCircleOutlined />,
				title: t(`project:menu.confirmDelete.title`),
				content: t(`project:menu.confirmDelete.content`),
				okText: t(`project:menu.confirmDelete.ok`),
				cancelText: t(`project:menu.confirmDelete.cancel`),
				onOk() {
					dispatch(deleteProject(projectId))
				},
			})
		},
		[dispatch, projectId, t]
	)

	const resolvedMenu = useMemo(
		() => (
			<Menu>
				{hasPrivilege(EDIT_PROJECT) && (
					<Menu.Item key={1}>
						<Styled.MenuButton onClick={_onEditProjectClick}> {t('project:menu.edit')}</Styled.MenuButton>
					</Menu.Item>
				)}
				{hasPrivilege(EDIT_PROJECT_EXPORT) && (
					<Menu.Item key={2}>
						<Styled.MenuButton onClick={_onExportProjectClick}> {t('project:menu.export')}</Styled.MenuButton>
					</Menu.Item>
				)}
				{hasPrivilege(DELETE_PROJECT) && (
					<Menu.Item key={3}>
						<Styled.MenuButton onClick={_onDeleteProjectClick}> {t('project:menu.delete')}</Styled.MenuButton>
					</Menu.Item>
				)}
			</Menu>
		),
		[hasPrivilege, _onEditProjectClick, _onExportProjectClick, _onDeleteProjectClick, t]
	)

	const _onItemClick = useCallback((itemId) => () => dispatch(navigateToItem(projectId, itemId)), [dispatch, projectId])

	const tasksColumns = [
		{ title: t('project:stats.table.name'), dataIndex: 'value', fixed: 'left' },
		{
			title: <Styled.TextRight>{t('project:stats.table.annotationPourcent')}</Styled.TextRight>,
			dataIndex: 'annotationPourcent',
			width: 125,
			render: (data) => (
				<Styled.TextRight>{isNumber(data) && t('root:units.percentLabel', { count: data })}</Styled.TextRight>
			),
		},
		{
			title: <Styled.TextRight>{t('project:stats.table.annotationCount')}</Styled.TextRight>,
			dataIndex: 'annotationCount',
			width: 125,
			render: (data) => <Styled.TextRight>{data}</Styled.TextRight>,
		},
		{
			title: <Styled.TextRight>{t('project:stats.table.velocity')}</Styled.TextRight>,
			dataIndex: 'velocity',
			width: 125,
			render: (data) => <Styled.TextRight>{data && t('root:units.secondLabel', { count: data })}</Styled.TextRight>,
		},
	]

	const itemsColumns = useMemo(
		() => [
			{
				title: t(`project:stats.table.name`),
				fixed: 'left',
				render: ({ _id, body }) => (
					<Styled.Link to={`/project/${projectId}/annotation/${_id}`} onClick={_onItemClick(_id)}>
						{!body || isHTML(body) ? `#...${_id.substring(_id.length - 6, _id.length)}` : `${body.substring(0, 30)}...`}
					</Styled.Link>
				),
			},
			{
				title: <Styled.TextRight>{t('project:stats.table.logCount')}</Styled.TextRight>,
				dataIndex: 'logCount',
				width: 125,
				render: (data) => <Styled.TextRight>{data}</Styled.TextRight>,
			},
			{
				title: <Styled.TextRight>{t('project:stats.table.velocity')}</Styled.TextRight>,
				dataIndex: 'velocity',
				width: 125,
				render: (data) => <Styled.TextRight>{data && t('root:units.secondLabel', { count: data })}</Styled.TextRight>,
			},
			{
				title: t('project:stats.table.annotated'),
				dataIndex: 'annotated',
				width: 125,
				render: (data) => (data ? t('project:stats.table.isAnnotated') : t('project:stats.table.isNotAnnotated')),
			},
			{
				title: t('project:stats.table.annotationValues'),
				dataIndex: 'annotationValues',
				width: 100,
				render: (data) => (!isEmpty(data) && data.length > 3 ? `${data.slice(0, 3).toString()}...` : data.toString()),
			},
			{
				title: t('project:stats.table.annotatedAt'),
				dataIndex: 'annotatedAt',
				width: 200,
				render: (date) => date && dayjs(date).fromNow(),
			},
		],
		[projectId, _onItemClick, t]
	)

	const tabList = useMemo(
		() => [
			{
				key: ITEMS,
				tab: t(`project:stats.table.${totalItems < 2 ? 'totalItems' : 'totalPluralItems'}`, {
					count: totalItems,
				}),
			},
			{
				key: TASKS,
				tab: t(`project:stats.table.${totalTasks < 2 ? 'totalTask' : 'totalPluralTasks'}`, {
					count: totalTasks,
				}),
			},
		],
		[t, totalTasks, totalItems]
	)

	const _onGuideClick = useCallback(() => dispatch(openGuideModal()), [dispatch])

	useEffect(
		() =>
			setHeaderActions &&
			setHeaderActions(
				<>
					<Styled.StartAnnotationButton to={`/project/${projectId}/annotation`} className="ant-btn ant-btn-primary">
						{t('project:annotation.start')}
					</Styled.StartAnnotationButton>
					{(hasPrivilege(EDIT_PROJECT) || hasPrivilege(EDIT_PROJECT_EXPORT) || hasPrivilege(DELETE_PROJECT)) && (
						<Dropdown overlay={resolvedMenu}>
							<Styled.MoreOutlined />
						</Dropdown>
					)}
					<Styled.Divider type={'vertical'} />
					<Styled.QuestionCircleOutlined onClick={_onGuideClick} />
				</>
			),
		[projectId, _onGuideClick, hasPrivilege, resolvedMenu, setHeaderActions, t]
	)

	const _onLoadingMoreClick = useCallback(
		() =>
			selectedStatsDisplay === ITEMS
				? dispatch(fetchProjectItems(projectId, itemsIndex + 1, PROJECT_ITEMS_SIZE))
				: dispatch(fetchProjectStatsTasks(projectId, tasksIndex + 1, PROJECT_STATS_TASKS_SIZE)),
		[selectedStatsDisplay, itemsIndex, tasksIndex, projectId, dispatch]
	)

	const _onEditFilterClick = useCallback(() => dispatch(openFilterModal()), [dispatch])

	const _onLogsFiltersChange = useCallback(
		(param) => dispatch(fetchProjectLogs(projectId, 0, PROJECT_LOGS_SIZE, param)),
		[dispatch, projectId]
	)

	const _onLoadingMore = useCallback(
		(type) => dispatch(fetchProjectLogs(projectId, logsIndex + 1, PROJECT_LOGS_SIZE, type)),
		[dispatch, projectId, logsIndex]
	)

	const _onTabChange = (key) => {
		if (!!key) {
			setSelectedStatsDisplay(key)
		}
	}

	return (
		<Page id={'homeProject'}>
			<Styled.Root>
				<Row gutter={['12', '12']}>
					<Col md={14} sm={24}>
						<Styled.LeftContainer>
							<Styled.ActionsContainer>
								<Styled.FunnelPlotOutlinedIcon />
								{filter?.total > 0 && <Styled.Badge count={`x${filter?.total}`} />}
								<Styled.EditButton type="link" onClick={_onEditFilterClick}>
									{t(`project:filter.editButton`)}
								</Styled.EditButton>
							</Styled.ActionsContainer>
							<Card
								activeTabKey={selectedStatsDisplay} //passer par le state activeTabKey /
								onTabChange={_onTabChange} //au click, setActiveTabeKey, setSelectedStatsDisplay
								isContentFullSize={true}
								tabList={tabList}
								tabProps={{ size: 'small' }}
							>
								{!isEmpty(selectedStatsDisplay === ITEMS ? itemsData : resolveTasksByCategories) ? (
									<Styled.Table
										expandIcon={() => null}
										expandRowByClick={true}
										defaultExpandAllRows={true}
										columns={selectedStatsDisplay === ITEMS ? itemsColumns : tasksColumns}
										dataSource={selectedStatsDisplay === ITEMS ? itemsData : resolveTasksByCategories}
										footer={
											(selectedStatsDisplay === ITEMS ? totalItems !== items.length : totalTasks !== tasks.length) &&
											(() => <Button onClick={_onLoadingMoreClick}>{t(`project:stats.table.loadMore`)}</Button>)
										}
									/>
								) : (
									<WarningMessage message={t(`project:warning.${!!filter?.total ? 'filter' : 'dataset'}`)} />
								)}
							</Card>
						</Styled.LeftContainer>
					</Col>
					<Col md={10} sm={24}>
						<Styled.RightContainer>
							<Card title={t('project:annotators.title')}>
								<AnnotatorsContainer annotators={annotators} currentUser={currentUser} />
							</Card>
							<Card title={t('project:logs.title')}>
								<LogsContainer
									logs={logs?.data || []}
									isProjectContext={true}
									onFiltersChange={_onLogsFiltersChange}
									onLoadingMore={_onLoadingMore}
									hasMoreLogs={totalLogs > logs?.data.length}
								/>
							</Card>
						</Styled.RightContainer>
					</Col>
				</Row>
			</Styled.Root>
		</Page>
	)
}

export default HomeProjectPage
