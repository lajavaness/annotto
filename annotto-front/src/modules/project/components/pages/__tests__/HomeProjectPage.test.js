import { LocationProvider } from '@reach/router'
import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components'
import { createStore } from 'redux-dynamic-modules'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import HomeProjectPage from 'modules/project/components/pages/HomeProjectPage'

const initialState = {}
const getStore = (state) => createStore({ initialState: state })
const getInstance = (state = initialState) => (
	<Provider store={getStore(state)}>
		<LocationProvider>
			<ThemeProvider theme={theme}>
				<HomeProjectPage />
			</ThemeProvider>
		</LocationProvider>
	</Provider>
)

describe('HomeProjectPage', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})
})
