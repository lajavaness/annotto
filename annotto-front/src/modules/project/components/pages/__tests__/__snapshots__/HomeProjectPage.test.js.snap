// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeProjectPage matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="sc-exiMOr QQPka"
    data-testid="__homeProject-page__"
  >
    <div
      class="sc-fxNMLY bfsYRs"
    >
      <div
        class="ant-row sc-eisiLS fdBAeg"
        style="margin: -6px -6px -6px -6px;"
      >
        <div
          class="ant-col sc-dkmKIT bZSqPx ant-col-sm-24 ant-col-md-14"
          style="padding: 6px 6px 6px 6px;"
        >
          <div
            class="sc-gGTEWt giHKah"
          >
            <div
              class="sc-dcwqsv cvvJZG"
            >
              <span
                aria-label="funnel-plot"
                class="anticon anticon-funnel-plot sc-hFXoIY iDvBoR"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="funnel-plot"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M880.1 154H143.9c-24.5 0-39.8 26.7-27.5 48L349 607.4V838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V607.4L907.7 202c12.2-21.3-3.1-48-27.6-48zM603.4 798H420.6V650h182.9v148zm9.6-226.6l-8.4 14.6H419.3l-8.4-14.6L334.4 438h355.2L613 571.4zM726.3 374H297.7l-85-148h598.6l-85 148z"
                  />
                </svg>
              </span>
              <button
                class="ant-btn ant-btn-link sc-iuAqRD lomPst"
                type="button"
              >
                <span>
                  project:filter.editButton
                </span>
              </button>
            </div>
            <div
              class="ant-card ant-card-contain-tabs sc-cvJIAo ivfBQH"
              data-testid="__card__"
            >
              <div
                class="ant-card-head"
              >
                <div
                  class="ant-card-head-wrapper"
                />
                <div
                  class="ant-tabs ant-tabs-top ant-tabs-small ant-card-head-tabs"
                >
                  <div
                    class="ant-tabs-nav"
                    role="tablist"
                  >
                    <div
                      class="ant-tabs-nav-wrap"
                    >
                      <div
                        class="ant-tabs-nav-list"
                        style="transform: translate(0px, 0px);"
                      >
                        <div
                          class="ant-tabs-tab ant-tabs-tab-active"
                        >
                          <div
                            aria-controls="rc-tabs-test-panel-items"
                            aria-selected="true"
                            class="ant-tabs-tab-btn"
                            id="rc-tabs-test-tab-items"
                            role="tab"
                            tabindex="0"
                          >
                            project:stats.table.totalItems
                          </div>
                        </div>
                        <div
                          class="ant-tabs-tab"
                        >
                          <div
                            aria-controls="rc-tabs-test-panel-tasks"
                            aria-selected="false"
                            class="ant-tabs-tab-btn"
                            id="rc-tabs-test-tab-tasks"
                            role="tab"
                            tabindex="0"
                          >
                            project:stats.table.totalTask
                          </div>
                        </div>
                        <div
                          class="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
                        />
                      </div>
                    </div>
                    <div
                      class="ant-tabs-nav-operations ant-tabs-nav-operations-hidden"
                    >
                      <button
                        aria-controls="rc-tabs-test-more-popup"
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-hidden="true"
                        class="ant-tabs-nav-more"
                        id="rc-tabs-test-more"
                        style="visibility: hidden; order: 1;"
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          aria-label="ellipsis"
                          class="anticon anticon-ellipsis"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="ellipsis"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                            />
                          </svg>
                        </span>
                      </button>
                    </div>
                  </div>
                  <div
                    class="ant-tabs-content-holder"
                  >
                    <div
                      class="ant-tabs-content ant-tabs-content-top"
                    >
                      <div
                        aria-hidden="false"
                        aria-labelledby="rc-tabs-test-tab-items"
                        class="ant-tabs-tabpane ant-tabs-tabpane-active"
                        id="rc-tabs-test-panel-items"
                        role="tabpanel"
                        tabindex="0"
                      />
                      <div
                        aria-hidden="true"
                        aria-labelledby="rc-tabs-test-tab-tasks"
                        class="ant-tabs-tabpane"
                        id="rc-tabs-test-panel-tasks"
                        role="tabpanel"
                        style="display: none;"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-card-body"
              >
                <span
                  class="sc-dkaWRx ilMNBg"
                  data-testid="__warning-message__"
                >
                  <span
                    aria-label="exclamation-circle"
                    class="anticon anticon-exclamation-circle sc-cVksOY kLInhh"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="exclamation-circle"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                      />
                      <path
                        d="M464 688a48 48 0 1096 0 48 48 0 10-96 0zm24-112h48c4.4 0 8-3.6 8-8V296c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8z"
                      />
                    </svg>
                  </span>
                  project:warning.dataset
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-col sc-dkmKIT bZSqPx ant-col-sm-24 ant-col-md-10"
          style="padding: 6px 6px 6px 6px;"
        >
          <div
            class="sc-iumJeC gNbmHu"
          >
            <div
              class="ant-card sc-cvJIAo gqPFHG"
              data-testid="__card__"
            >
              <div
                class="ant-card-head"
              >
                <div
                  class="ant-card-head-wrapper"
                >
                  <div
                    class="ant-card-head-title"
                  >
                    project:annotators.title
                  </div>
                </div>
              </div>
              <div
                class="ant-card-body"
              >
                <div
                  class="sc-gsBqHK gsESNF"
                >
                  <div
                    class="sc-kmATbt cDljgH"
                  >
                    project:annotators.noAnnotators
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-card sc-cvJIAo gqPFHG"
              data-testid="__card__"
            >
              <div
                class="ant-card-head"
              >
                <div
                  class="ant-card-head-wrapper"
                >
                  <div
                    class="ant-card-head-title"
                  >
                    project:logs.title
                  </div>
                </div>
              </div>
              <div
                class="ant-card-body"
              >
                <div
                  class="sc-gIRjGK gPIBjz"
                >
                  <div
                    class="sc-gHfsNP bHajzS"
                  >
                    <div
                      class="ant-radio-group ant-radio-group-outline sc-dkAqVg jnIBqp"
                    >
                      <label
                        class="ant-radio-button-wrapper ant-radio-button-wrapper-checked"
                      >
                        <span
                          class="ant-radio-button ant-radio-button-checked"
                        >
                          <input
                            checked=""
                            class="ant-radio-button-input"
                            type="radio"
                            value=""
                          />
                          <span
                            class="ant-radio-button-inner"
                          />
                        </span>
                        <span>
                          project:logs.all
                        </span>
                      </label>
                      <label
                        class="ant-radio-button-wrapper"
                      >
                        <span
                          class="ant-radio-button"
                        >
                          <input
                            class="ant-radio-button-input"
                            type="radio"
                            value="tags"
                          />
                          <span
                            class="ant-radio-button-inner"
                          />
                        </span>
                        <span>
                          project:logs.tags
                        </span>
                      </label>
                      <label
                        class="ant-radio-button-wrapper"
                      >
                        <span
                          class="ant-radio-button"
                        >
                          <input
                            class="ant-radio-button-input"
                            type="radio"
                            value="comment"
                          />
                          <span
                            class="ant-radio-button-inner"
                          />
                        </span>
                        <span>
                          project:logs.comment
                        </span>
                      </label>
                      <label
                        class="ant-radio-button-wrapper"
                      >
                        <span
                          class="ant-radio-button"
                        >
                          <input
                            class="ant-radio-button-input"
                            type="radio"
                            value="config"
                          />
                          <span
                            class="ant-radio-button-inner"
                          />
                        </span>
                        <span>
                          project:logs.config
                        </span>
                      </label>
                      <label
                        class="ant-radio-button-wrapper"
                      >
                        <span
                          class="ant-radio-button"
                        >
                          <input
                            class="ant-radio-button-input"
                            type="radio"
                            value="annotation"
                          />
                          <span
                            class="ant-radio-button-inner"
                          />
                        </span>
                        <span>
                          project:logs.annotation
                        </span>
                      </label>
                    </div>
                    <div
                      class="sc-hkwmXC hRFDcp"
                    />
                  </div>
                  <div
                    class="ant-divider ant-divider-horizontal sc-cHjwLt btQRTV"
                    role="separator"
                  />
                  <form
                    class="ant-form ant-form-horizontal sc-bSFUlv hxSKxA"
                  >
                    <div
                      class="ant-row ant-form-item sc-jvfqPk cqriyh"
                    >
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <span
                              class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
                            >
                              <textarea
                                class="ant-input"
                                id="comment"
                                placeholder="project:comments.placeholder"
                                style="height: 0px; min-height: -9007199254740991px; max-height: 9007199254740991px; overflow-y: hidden; resize: none; overflow-x: hidden;"
                              />
                              <span
                                aria-label="close-circle"
                                class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                                role="button"
                                tabindex="-1"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="close-circle"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ant-row ant-form-item sc-jvfqPk cqriyh"
                    >
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <button
                              class="ant-btn ant-btn-default ant-btn-icon-only sc-edoYdd jPhrgK"
                              type="submit"
                            >
                              <span
                                aria-label="send"
                                class="anticon anticon-send"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="send"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <path
                                    d="M931.4 498.9L94.9 79.5c-3.4-1.7-7.3-2.1-11-1.2a15.99 15.99 0 00-11.7 19.3l86.2 352.2c1.3 5.3 5.2 9.6 10.4 11.3l147.7 50.7-147.6 50.7c-5.2 1.8-9.1 6-10.3 11.3L72.2 926.5c-.9 3.7-.5 7.6 1.2 10.9 3.9 7.9 13.5 11.1 21.5 7.2l836.5-417c3.1-1.5 5.6-4.1 7.2-7.1 3.9-8 .7-17.6-7.2-21.6zM170.8 826.3l50.3-205.6 295.2-101.3c2.3-.8 4.2-2.6 5-5 1.4-4.2-.8-8.7-5-10.2L221.1 403 171 198.2l628 314.9-628.2 313.2z"
                                  />
                                </svg>
                              </span>
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
